src/sentry/search/django/backend.py

```diff
@@ -195,6 +197,11 @@ def query(self, project, tags=None, environment=None, sort_by='date', limit=100,
         if tags is None:
             tags = {}
 
+        # Any EMPTY value means there can be no results for this query so
+        # return an empty paginator immediately.
+        if any(val == EMPTY for _, val in six.iteritems(tags)):
+            return Paginator(Group.objects.none()).get_result()
```

Comment:
Moved this up to the top because @tkaemming said `EMPTY` always means nothing? I honestly still need to go dig up the stack to see where this comes from, because I don't understand why we allow users to initiate a search that will never return results?

But, I moved it up because each tagstore backend was manually handling this and I'd rather do it before.

Reply:
> Moved this up to the top because @tkaemming said EMPTY always means no results (and existing impls seem to agree)?

Yeah, assuming we're using the existing tagstore and search as the reference implementation, that's my understanding as well.

> I honestly still need to go dig up the stack to see where this comes from, because I don't understand why we allow users to initiate a search that will never return results?

Saving you a `grep`, it's returned here:
https://github.com/getsentry/sentry/blob/d2791de0c85c130bfa48549dc5538d8afe2f677b/src/sentry/search/utils.py#L20-L33

Sidenote: We should probably actually do the TODO here because the results are wrong for the new environments features. It seems like this should be encapsulated in the search backend anyway, rather than the `parse_query` method?

Which is called here:
https://github.com/getsentry/sentry/blob/d2791de0c85c130bfa48549dc5538d8afe2f677b/src/sentry/search/utils.py#L366-L369

> But, I moved it up because each tagstore backend was manually handling this and I'd rather do it before.

Agree 100%.

Reply:
>  # TODO(dcramer): add environment support

Going to handle this next/now in a separate PR.

Reply:
https://github.com/getsentry/sentry/pull/8054